### PR TITLE
Enable devlocal auth from flag and use environment checking

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

To support Load Testing in #1597 we need a way to turn on Devlocal Auth in experimental. This adds a flag that enables the devlocal auth but will check against the deployed environment and server names and fail if the environment is wrong. 

## Reviewer Notes

Could this be made more safe so we don't enable devlocal auth in production or staging? I am checking devlocal flag against:

- The Environment flag (`--environment`)
- All the Server Name Flags (my, office, tsp, and admin)

The combination here means multiple things need to fail in order to see this deployed in prod/staging.

## Setup

Run the app. Try changing your `.envrc.local` vars for `DEVLOCAL_AUTH`.

## Code Review Verification Steps

* [x] Request review from a member of a different team.
* [x] Have the Pivotal acceptance criteria been met for this change?

### CODE REVIEW

**Comment by chrisgilmerproj** [comment]
Can I couple this to the hostname check? Not just dbEnv but also hostname and maybe another thing? The stronger the coupling the less likely this will happen.

**Comment by chrisgilmerproj** [comment]
@pjdufour-truss - I'm pretty sure this is done but you're the person I rely on for security checks. Is there another way to get the app deployed in AWS with devlocal auth enabled that doesn't possibly compromise prod/staging?

File: pkg/cli/devlocal.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+// InitDevlocalFlags initializes the Devlocal command line flags
+func InitDevlocalFlags(flag *pflag.FlagSet) {
+	flag.Bool(DevlocalAuthFlag, false, "Enable the devlocal auth system for logging in without Login.gov.")
+}
+
+// CheckDevlocal validates the Devlocal command line flags
+func CheckDevlocal(v *viper.Viper) error {
+	if devlocalAuthEnabled := v.GetBool(DevlocalAuthFlag); devlocalAuthEnabled {
```
Comment (pjdufour-truss) [inline pkg/cli/devlocal.go]: Declaring a variable seems to be redundant here.

File: pkg/cli/dbconn.go [hunk 2]
```diff
@@ -31,18 +31,59 @@ const (
...
+
+	// DbEnvContainer is the Container DB Env name
+	DbEnvContainer string = "container"
+	// DbEnvTest is the Test DB Env name
+	DbEnvTest string = "test"
+	// DbEnvDevelopment is the Development DB Env name
+	DbEnvDevelopment string = "development"
+
+	// SSLModeDisable is the disable SSL Mode
+	SSLModeDisable string = "disable"
```
Comment (pjdufour-truss) [inline pkg/cli/dbconn.go]: I really like expanding our use of const values, but could we split the database ssl modes into a separate PR and jam it through without waiting on this bigger operational change?

File: scripts/run-e2e-test-docker [hunk 3]
```diff
@@ -13,6 +13,8 @@ CONTAINER_CYPRESS=cypress
 MILMOVEHOST=milmovelocal
 OFFICEHOST=officelocal
 TSPHOST=tsplocal
+ADMINHOST=adminlocal
```
Comment (pjdufour-truss) [inline scripts/run-e2e-test-docker]: These look like fixes for the admin app.  Could we split these into their own PR?

File: pkg/cli/dbconn.go (hunk 2 above)
Comment (chrisgilmerproj) [reply pkg/cli/dbconn.go]: I've moved it here: https://github.com/transcom/mymove/pull/2121

File: scripts/run-e2e-test-docker (hunk 3 above)
Comment (chrisgilmerproj) [reply scripts/run-e2e-test-docker]: I've moved it here: https://github.com/transcom/mymove/pull/2122. The reason I added these is that I'm checking the config against the admin hosts and that made it fail in the e2e tests. It's ok to have it separated though as it should have been added by the Admin team along with e2e tests.

File: pkg/cli/devlocal.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+func CheckDevlocal(v *viper.Viper) error {
+	if v.GetBool(DevlocalAuthFlag) {
+		// Check against the Environment
+		allowedEnvironments := []string{
+			EnvironmentDevelopment,
+			EnvironmentTest,
+			EnvironmentExperimental,
+		}
+		if environment := v.GetString(EnvironmentFlag); !stringSliceContains(allowedEnvironments, environment) {
+			return errors.Errorf("Devlocal Auth cannot run in the '%s' environment, only in %v", environment, allowedEnvironments)
```
Comment (jim) [inline pkg/cli/devlocal.go]: I wonder if we want to `panic` here so that the returned error can't be ignored.

Alternatively, we could enforce error checking somewhere in the build process.

Comment (chrisgilmerproj) [reply pkg/cli/devlocal.go]: We do error checking in the `cmd/milmove/main.go` and if any checks fail the app won't start up. So this is ok here and we can deal with the errors in the main function I think.

Comment (jim) [reply pkg/cli/devlocal.go]: My worry was that if we forget that error check in a future change, all of the care put into crafting these checks would be for naught. But I'm OK with leaving this as is for now.

**Comment by chrisgilmerproj** [comment]
> I find the nomenclature fairly confusing; what is the difference between `ENV` and `ENVIRONMENT`? My worry is that this could lead to confusion or mistakes in the future. Can we rename one of these, perhaps `ENVIRONMENT` since I think it's actually the specific cluster that we're deployed to?

You're not alone @jim. The problem is that `ENV` was always meant to be connected to the Database and how Pop does work. Digging through the old commits showed me that `ENV` got co-opted to also mean the level we used for Logging. I've opted to not change `ENV` at all (yet) but my hope is to make one called `--db-env` and one called `--logging-env` and keep them separate.

`ENVIRONMENT` refers to the deployed environment which refers to one of `development`, `testing`, `experimental`, `staging`, or `prod`. It does have an effect in local development so it's difficult to just say that it refers to the AWS clusters.  For my part I think `ENVIRONMENT` should be the one that stays as is and `ENV` ought to be split out like I mentioned above.

I'd like to push any work of renaming the flags outside of this PR. There's a bunch of little things I'd have to fix and I think for getting the devlocal flag in place this work is sufficient.

File: pkg/cli/devlocal.go (hunk 4 above)
Comment (chrisgilmerproj) [reply pkg/cli/devlocal.go]: How about adding a panic here: https://github.com/transcom/mymove/pull/2081/files#diff-846b2b1c0f3cb7d032edae2ebf248001R176

Its not exactly how the rest of the code execution works but it would allow us to capture all 5 of these checks in one place.

**Comment by chrisgilmerproj** [comment]
I've deployed to experimental and it worked! Nobody will see the `Local Sign-in` button like in development. You have to browse directly to `/devlocal-auth/login` to get the login.

- https://my.experimental.move.mil/devlocal-auth/login
- https://office.experimental.move.mil/devlocal-auth/login
- https://tsp.experimental.move.mil/devlocal-auth/login

I'll disable this in experimental and enable it only in the load testing PRs.
